Replace `assert!` with `debug_assert!` and add a new profile `release-debug`

silathdiir opened the pull request:
Close https://github.com/appliedzkp/zkevm-circuits/issues/288

### Summary

1. Only replace `assert!` with `debug_assert!` in no-test code.
2. Add a new profile `release-debug` inherited from `release` profile.

Reference comment https://github.com/appliedzkp/zkevm-circuits/issues/288#issuecomment-1015413048, I set `debug = true` for debug symbols. And `overflow-checks = true` for enabling integer overflow checks, `incremental = true` for incremental compilation.

It seems that the profile name `debug` is reserved. An error occurred when adding a custom profile named `debug` as below.

```
error: failed to parse manifest at `/zkevm-circuits/Cargo.toml`

Caused by:
  profile name `debug` is reserved
  To configure the default development profile, use the name `dev` as in [profile.dev]
  See https://doc.rust-lang.org/cargo/reference/profiles.html for more on configuring profiles.
```

lispc commented:
I think 'debug_assertion=true' is enough to enable debug_assert.   So what is reason for add 'debug = true' and 'incremental = true'? For stack trace and linking speed? You can add more information in PR comment

silathdiir commented:
> I think 'debug_assertion=true' is enough to enable debug_assert. So what is reason for add 'debug = true' and 'incremental = true'? For stack trace and linking speed? You can add more information in PR comment

It seems that the comment https://github.com/appliedzkp/zkevm-circuits/issues/288#issuecomment-1015413048 requires the debug symbols, so I add `debug = true`.

I updated the PR comment for `debug = true`, `overflow-checks = true` and `incremental = true`. Thanks.

And I also fixed the failed test, there was code of HashMap insert in `assert!` (fixed to `debug_assert!`). I move the `insert` out of `debug_assert!`.

scroll-dev commented:
moved to https://github.com/appliedzkp/zkevm-circuits/pull/310
